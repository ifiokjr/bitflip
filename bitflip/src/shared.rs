pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
