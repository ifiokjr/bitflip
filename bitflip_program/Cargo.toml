[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
