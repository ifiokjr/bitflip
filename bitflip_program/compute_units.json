{
  "instructions": {
    "config_initialize": {
      "compute_units": 17781,
      "rounded_compute_units": 20000,
      "description": "Initialize the config"
    },
    "config_update_authority": {
      "compute_units": 2464,
      "rounded_compute_units": 10000,
      "description": "Update the authority of the config"
    },
    "flip_bit_test": {
      "compute_units": 65231,
      "rounded_compute_units": 70000,
      "description": "Flip a single bit from `0` to `1`"
    },
    "game_initialize": {
      "compute_units": 13227,
      "rounded_compute_units": 20000,
      "description": "Initialize the game"
    },
    "game_refresh_signer": {
      "compute_units": 2371,
      "rounded_compute_units": 10000,
      "description": "Refresh the signer"
    },
    "token_group_initialize": {
      "compute_units": 47820,
      "rounded_compute_units": 50000,
      "description": "Initialize the token group"
    },
    "token_initialize:bit": {
      "compute_units": 63507,
      "rounded_compute_units": 70000,
      "description": "Initialize group token: Bit"
    },
    "token_initialize:gibibit": {
      "compute_units": 60069,
      "rounded_compute_units": 70000,
      "description": "Initialize member token: GibiBit"
    },
    "token_initialize:kibibit": {
      "compute_units": 60068,
      "rounded_compute_units": 70000,
      "description": "Initialize member token: KibiBit"
    },
    "token_initialize:mebibit": {
      "compute_units": 60065,
      "rounded_compute_units": 70000,
      "description": "Initialize member token: MebiBit"
    }
  },
  "version": "0.0.0"
}