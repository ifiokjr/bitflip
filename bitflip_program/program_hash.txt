50bd03d71a9fee65661b2bf98544cd5892ab8ff4ac6d8e5e1cfc186a2a7f6fa7
