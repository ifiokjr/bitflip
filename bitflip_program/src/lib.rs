mod constants;
pub mod cpi;
mod errors;
mod events;
#[cfg(feature = "client")]
mod instructions;
mod loaders;
mod pda;
mod processor;
mod state;
mod utils;

use steel::*;

pub use crate::constants::*;
pub use crate::errors::*;
pub use crate::events::*;
#[cfg(feature = "client")]
pub use crate::instructions::*;
pub use crate::loaders::*;
pub use crate::pda::*;
pub use crate::processor::*;
pub use crate::state::*;
pub use crate::utils::*;

solana_package_metadata::declare_id_with_package_metadata!("solana.program-id");

#[cfg(not(feature = "no-entrypoint"))]
entrypoint!(process_instruction);

// #[cfg(not(feature = "no-entrypoint"))]
solana_security_txt::security_txt! {
	// Required fields
	name: "Bitflip",
	project_url: "https://bitflip.art",
	contacts: "link:https://github.com/ifiokjr/bitflip/security/advisories/new,email:security@kickjump.co,discord:https://bitflip.art/discord",
	policy: "https://github.com/ifiokjr/bitflip/blob/main/security.md",

	// Optional Fields
	preferred_languages: "en",
	source_code: "https://github.com/ifiokjr/bitflip/tree/main/bitflip_program/",
	source_revision: default_env::default_env!("GITHUB_SHA", ""),
	source_release: default_env::default_env!("GITHUB_REF_NAME", ""),
	auditors: concat!("Verifier pubkey: ", default_env::default_env!("GITHUB_SHA", "")),
	encryption: "",
	acknowledgements: "Thank you to our bug bounty degens!"
}

#[cfg(test)]
pub(crate) fn leak<T>(value: T) -> &'static mut T {
	Box::leak(Box::new(value))
}
