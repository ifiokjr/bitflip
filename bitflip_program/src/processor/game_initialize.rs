use steel::*;

use super::BitflipInstruction;
use crate::BitflipError;
use crate::ConfigState;
use crate::GameState;
use crate::ID;
use crate::SEED_GAME;
use crate::SEED_PREFIX;
use crate::create_pda_config;
use crate::get_pda_game;

pub fn process_game_initialize(accounts: &[AccountInfo]) -> ProgramResult {
	let [
		authority_info,
		access_signer_info,
		refresh_signer_info,
		config_info,
		game_info,
		system_program_info,
	] = accounts
	else {
		return Err(ProgramError::NotEnoughAccountKeys);
	};

	authority_info
		.is_signer()?
		.is_writable()?
		.has_owner(&system_program::ID)?;
	access_signer_info.is_empty()?.is_signer()?;
	refresh_signer_info.is_empty()?.is_signer()?;
	config_info.is_type::<ConfigState>(&ID)?;
	game_info.is_empty()?.is_writable()?;
	system_program_info.is_program(&system_program::ID)?;

	let config = config_info.as_account::<ConfigState>(&ID)?;
	let config_key = create_pda_config(config.bump)?;
	let (game_key, game_bump) = get_pda_game(config.game_index);

	if authority_info.key.ne(&config.authority) {
		return Err(BitflipError::Unauthorized.into());
	}

	if config_info.key.ne(&config_key) || game_info.key.ne(&game_key) {
		return Err(ProgramError::InvalidSeeds);
	}

	// create the onchain account
	create_account_with_bump::<GameState>(
		game_info,
		system_program_info,
		authority_info,
		&ID,
		&[SEED_PREFIX, SEED_GAME, &config.game_index.to_le_bytes()],
		game_bump,
	)?;

	let game = game_info.as_account_mut::<GameState>(&ID)?;
	*game = GameState::new(
		*access_signer_info.key,
		*refresh_signer_info.key,
		config.game_index,
		game_bump,
	);

	Ok(())
}

#[repr(C)]
#[derive(Clone, Copy, Debug, Pod, Zeroable)]
pub struct GameInitialize {}

instruction!(BitflipInstruction, GameInitialize);

#[cfg(test)]
mod tests {
	use std::cell::RefCell;
	use std::rc::Rc;

	use assert2::check;
	use solana_sdk::native_loader;

	use super::*;
	use crate::get_pda_config;
	use crate::leak;

	#[test_log::test]
	fn should_have_enough_accounts() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		let result = process_game_initialize(&accounts[..5]);
		check!(result.unwrap_err() == ProgramError::NotEnoughAccountKeys);

		Ok(())
	}

	#[test_log::test]
	fn authority_should_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[0];
		authority_info.is_signer = false;

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn authority_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[0];
		authority_info.is_writable = false;

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn authority_should_be_from_config() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[0];
		authority_info.key = leak(Pubkey::new_unique());

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::Custom(BitflipError::Unauthorized.into()));

		Ok(())
	}

	#[test_log::test]
	fn access_signer_should_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let access_signer_info = &mut accounts[1];
		access_signer_info.is_signer = false;

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn refresh_signer_should_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let refresh_signer_info = &mut accounts[2];
		refresh_signer_info.is_signer = false;

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn config_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[3];
		config_info.key = leak(Pubkey::new_unique());

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn config_should_have_valid_data() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[3];
		config_info.data = Rc::new(RefCell::new(leak(vec![1u8; 8])));

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidAccountData);

		Ok(())
	}

	#[test_log::test]
	fn config_should_have_valid_owner() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[3];
		config_info.owner = leak(Pubkey::new_unique());

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidAccountOwner);

		Ok(())
	}

	#[test_log::test]
	fn game_should_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let game_info = &mut accounts[4];
		game_info.key = leak(Pubkey::new_unique());

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn game_should_be_empty() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let game_info = &mut accounts[4];
		game_info.data = Rc::new(RefCell::new(leak(vec![1, 0, 0, 0, 0, 0, 0, 0])));

		let result = process_game_initialize(&accounts);
		check!(result.unwrap_err() == ProgramError::AccountAlreadyInitialized);

		Ok(())
	}

	#[test_log::test]
	fn game_should_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let game_info = &mut accounts[4];
		game_info.is_writable = false;

		Ok(())
	}

	fn create_account_infos<'info>() -> [AccountInfo<'info>; 6] {
		let authority = Pubkey::new_unique();
		let authority_key = leak(authority);
		let authority_lamports = leak(1_000_000_000);
		let authority_data = leak(vec![]);
		let access_signer_key = leak(Pubkey::new_unique());
		let access_signer_lamports = leak(0);
		let access_signer_data = leak(vec![]);
		let refresh_signer_key = leak(Pubkey::new_unique());
		let refresh_signer_lamports = leak(0);
		let refresh_signer_data = leak(vec![]);
		let config_key = leak(get_pda_config().0);
		let config_lamports = leak(0);
		let config_data = {
			let config_bump = get_pda_config().1;
			let mut data = vec![0u8; 8];
			data[0] = ConfigState::discriminator();
			data.append(
				&mut ConfigState::new(
					authority,
					config_bump,
					u8::MAX,
					u8::MAX,
					u8::MAX,
					u8::MAX,
					u8::MAX,
				)
				.to_bytes()
				.to_vec(),
			);
			leak(data)
		};
		let game_key = leak(get_pda_game(0).0);
		let game_lamports = leak(0);
		let game_data = leak(vec![]);
		let system_program_lamports = leak(1_000_000_000);
		let system_program_data = leak(vec![]);

		let authority_info = AccountInfo::new(
			authority_key,
			true,
			true,
			authority_lamports,
			authority_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let access_signer_info = AccountInfo::new(
			access_signer_key,
			true,
			false,
			access_signer_lamports,
			access_signer_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let refresh_signer_info = AccountInfo::new(
			refresh_signer_key,
			true,
			false,
			refresh_signer_lamports,
			refresh_signer_data,
			&system_program::ID,
			false,
			u64::MAX,
		);
		let config_info = AccountInfo::new(
			config_key,
			false,
			true,
			config_lamports,
			config_data,
			&ID,
			false,
			u64::MAX,
		);
		let game_info = AccountInfo::new(
			game_key,
			false,
			true,
			game_lamports,
			game_data,
			&ID,
			false,
			u64::MAX,
		);
		let system_program_info = AccountInfo::new(
			&system_program::ID,
			false,
			false,
			system_program_lamports,
			system_program_data,
			&native_loader::ID,
			true,
			u64::MAX,
		);

		[
			authority_info,
			access_signer_info,
			refresh_signer_info,
			config_info,
			game_info,
			system_program_info,
		]
	}
}
