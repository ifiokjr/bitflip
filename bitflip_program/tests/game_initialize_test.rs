#![cfg(feature = "client")]

use std::future::Future;

use assert2::check;
use bitflip_program::GameState;
use bitflip_program::game_initialize;
use bitflip_program::get_pda_game;
use shared::ToRpcClient;
use shared::create_authority_keypair;
use shared::create_config_accounts;
use shared::create_program_context_with_factory;
use shared::create_token_group_accounts;
use solana_sdk::signature::Keypair;
use solana_sdk::transaction::VersionedTransaction;
use steel::*;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;

mod shared;

#[test_log::test(tokio::test)]
async fn game_initialize_test() -> anyhow::Result<()> {
	shared_game_initialize_test(create_banks_client_rpc).await?;
	Ok(())
}

#[cfg(feature = "test_validator")]
#[test_log::test(tokio::test)]
async fn game_initialize_test_validator() -> anyhow::Result<()> {
	let compute_units = shared_game_initialize_test(create_validator_rpc).await?;
	let rounded_compute_units = bitflip_program::round_compute_units_up(compute_units);

	check!(rounded_compute_units == 20_000);
	insta::assert_snapshot!(format!("{rounded_compute_units} CU"));
	shared::save_compute_units("game_initialize", compute_units, "Initialize the game")?;
	Ok(())
}

async fn create_banks_client_rpc() -> anyhow::Result<impl ToRpcClient> {
	let provider = create_program_context_with_factory(|p| {
		let mut config_state_accounts = create_config_accounts();
		config_state_accounts.extend(create_token_group_accounts());

		for (key, account) in config_state_accounts {
			p.add_account(key, account.into());
		}
	})
	.await?;

	Ok(provider)
}

#[cfg(feature = "test_validator")]
async fn create_validator_rpc() -> anyhow::Result<impl ToRpcClient> {
	let mut accounts = create_config_accounts();
	accounts.extend(create_token_group_accounts());

	let runner = shared::create_runner_with_accounts(accounts).await;

	Ok(runner)
}

async fn shared_game_initialize_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
) -> anyhow::Result<u64> {
	let game_index = 0;
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let access_signer_keypair = Keypair::new();
	let refresh_signer_keypair = Keypair::new();
	let access_signer = access_signer_keypair.pubkey();
	let refresh_signer = refresh_signer_keypair.pubkey();
	let game = get_pda_game(game_index).0;
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let ix = game_initialize(&authority, &access_signer, &refresh_signer, game_index);
	let mut transaction =
		VersionedTransaction::new_unsigned_v0(&authority, &[ix], &[], recent_blockhash)?;

	let simulation = rpc.simulate_transaction(&transaction).await?;
	log::info!("simulation: {simulation:#?}");
	let compute_units = simulation.value.units_consumed.unwrap();

	transaction
		.try_sign(&[&authority_keypair], None)?
		.try_sign(&[&access_signer_keypair], None)?
		.try_sign(&[&refresh_signer_keypair], None)?;

	let signature = rpc.send_and_confirm_transaction(&transaction).await?;
	rpc.confirm_transaction(&signature).await?;

	let game_state_data = rpc.get_account_data(&game).await?;
	let game_state_account = GameState::try_from_bytes(&game_state_data)?;
	let access_signer_redaction = create_insta_redaction(access_signer, "access_signer:pubkey");
	let refresh_signer_redaction = create_insta_redaction(refresh_signer, "refresh_signer:pubkey");
	insta::assert_compact_json_snapshot!(game_state_account,{
		".accessSigner" => insta::dynamic_redaction(access_signer_redaction),
		".refreshSigner" => insta::dynamic_redaction(refresh_signer_redaction),

	}, @r#"
 {
   "refreshSigner": "[refresh_signer:pubkey]",
   "accessSigner": "[access_signer:pubkey]",
   "accessExpiry": 0,
   "startTime": 0,
   "gameIndex": 0,
   "sectionIndex": 0,
   "bump": 253
 }
 "#);

	let refresh_signer_lamports = rpc.get_balance(&refresh_signer).await?;
	log::info!("refresh_signer_lamports: {refresh_signer_lamports}");
	insta::assert_snapshot!(refresh_signer_lamports, @"5890880");

	Ok(compute_units)
}
