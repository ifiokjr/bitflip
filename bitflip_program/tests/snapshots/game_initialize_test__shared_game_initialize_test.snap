---
source: bitflip_program/tests/game_initialize_test.rs
expression: game_state_account
snapshot_kind: text
---
{
  "version": 0,
  "refreshSigner": "[refresh_signer:pubkey]",
  "accessSigner": "[access_signer:pubkey]",
  "accessExpiry": 0,
  "startTime": 0,
  "gameIndex": 0,
  "sectionIndex": 0,
  "bump": 253,
  "padding": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0
  ]
}
